skimage/draw/draw.py

```diff
@@ -520,7 +554,20 @@ def circle_perimeter(r, c, radius, method='bresenham', shape=None):
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
     """
-    return _circle_perimeter(r, c, radius, method, shape)
+    if not np.can_cast(r, np.integer) or not np.can_cast(c, np.integer):
+        raise TypeError("r and c must be integers")
+
+    rr, cc = _circle_perimeter(radius, method)
+    rr, cc = np.unique(np.stack([rr, cc]), axis=1)
+
+    rr += r
+    cc += c
+    if shape is not None:
+        keep = _inside_image(rr, cc, shape)
+        rr = rr[keep]
+        cc = cc[keep]
```

Comment:
This pattern is repeated over and over. If you're going to change the `_inside_image` logic (which seems unnecessary, but ok), perhaps we should let the function do the work of filtering as well? ie:

```python
rr, cc = _inside_image((rr, cc), shape)
(rr, cc), values = _inside_image((rr, cc), shape, values=values)
(pp, rr, cc), values = _inside_image((pp, rr, cc), shape, values=values)
```

If we want to avoid changing the return based on the input, we can always return values, returning None if none were provided.

Reply:
I have not idea ether. That's why I made a +1 on @lagru 's comment in the issue. At that time, I basically followed the implementation linked in the references.

Reply:
> This pattern is repeated over and over.

Yes but considering how stupidly simple that snippet is one could argue that the code is actually better. But that's not something I feel strongly about. :wink: